modules/installation-mirror-repository.adoc

```diff
@@ -57,11 +92,103 @@ $ oc adm -a ${LOCAL_SECRET_JSON} release mirror \
 ----
 +
 This command pulls the release information as a digest, and its output includes
-the `imageContentSources` data that you require when you install your cluster.
+text that resembles the following sample:
+
+. Record the
+ifdef::install[]
+`imageContentSources`
+endif::install[]
+ifdef::update[]
+`ImageContentSourcePolicy`
+endif::update[]
+section from the output of the previous
+command. This information is required
+ifdef::install[]
+during {product-title} installation.
+endif::install[]
+ifdef::update[]
+when you update your {product-title} cluster.
+endif::update[]
+
+. Build the signature for the content that you mirrored and verify that it is
+signed by an official Red Hat key:
+.. Set the digest for the version to
+ifdef::install[]
+install:
+endif::install[]
+ifdef::update[]
+update to:
+endif::update[]
++
+----
+$ DIGEST="$(oc adm release info quay.io/${PRODUCT_REPO}/${RELEASE_NAME}:${OCP_RELEASE} | sed -n 's/Pull From: .*@//p')" <1>
+----
+<1> For `<release_version>`, specify the version number of {product-title} to
+that you mirrored content for.
+
+.. Build a signature URI:
++
+----
+$ URI="https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/${DIGEST/:/=}/signature-<n>" <1>
+----
+<1> Specify the name of the signature to store. The signature name must be in the format of `signature-<n>`, where `<n>` is a number. If you have multiple signatures, the signature that includes the lowest number in its name is applied first, and other signatures are applied in increasing numerical order if one fails.
+
+.. Download the signature:
++
+----
+$ wget "${URI}"
+----
+
+.. Download the keys that were used to sign the signature:
++
+----
+$ wget https://www.redhat.com/security/data/f21541eb.txt
+$ wget https://www.redhat.com/security/data/fd431d51.txt
+----
+
+.. Import the keys:
++
+----
+$ gpg --no-default-keyring --keyring ./temp.keyring --import ./f21541eb.txt
+$ gpg --no-default-keyring --keyring ./temp.keyring --import ./fd431d51.txt
+----
+
+.. Verify the signature:
++
+----
+$ gpg --no-default-keyring --keyring ./temp.keyring --verify <signature_name> <1>
```

Comment:
Just noticed that this is checking for a valid sig, but not checking the signed content (so a single valid sig could be applied to multiple images).  I'll get instructions for checking content when I get in.

Reply:
Oh, so what we want is `--decrypt`, not `--verify`.

```console
$ gpg --no-default-keyring --keyring ./temp.keyring --decrypt signature-1 
{"critical": {"image": {"docker-manifest-digest": "sha256:a7e97365d16d8d920fedd3684b018b780337e069deb1dd8500e866c0d6110334"}, "type": "atomic container signature", "identity": {"docker-reference": "quay.io/openshift-release-dev/ocp-release:4.1.20"}}}gpg: Signature made Tue 15 Oct 2019 02:21:18 PM PDT using RSA key ID F21541EB
gpg: Good signature from "Red Hat, Inc. (beta key 2) <security@redhat.com>"
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: B08B 659E E86A F623 BC90  E8DB 938A 80CA F215 41EB
```

We want:

```console
$ gpg --no-default-keyring --keyring ./temp.keyring --decrypt signature-1 | jq -r '.critical.image["docker-manifest-digest"]'
gpg: Signature made Tue 15 Oct 2019 02:21:18 PM PDT using RSA key ID F21541EB
gpg: Good signature from "Red Hat, Inc. (beta key 2) <security@redhat.com>"
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: B08B 659E E86A F623 BC90  E8DB 938A 80CA F215 41EB
sha256:a7e97365d16d8d920fedd3684b018b780337e069deb1dd8500e866c0d6110334
```

to match `${DIGEST}`.  Docs on the full structure [here][1], altough I dunno how formal we want to get.  I'll work on `oc` tooling for this next week so I'm ok cutting some corners in the docs here in the meantime, as long as we're confirming that the expected digest is in the signed content.

[1]: https://github.com/containers/image/blob/8c5dd5b6e7c0253cf2f0f30c8ecd8d5e35cf03c2/docs/containers-signature.5.md#json-data-format

Reply:
Not even that.

This code, which resolves `${DIGEST}` from the `:${OCP_RELEASE}` tag above, and then uses `:${OCP_RELEASE}` again in `oc adm release mirror`, is **INSECURE** and **MUST NOT** be used; the registry can maliciously modify the `:${OCP_RELEASE}` tag in the meantime. (Alternatively, if you trust the registry never to do that, and to always point at legitimate OCP images, you don’t need to bother with signatures, by definition.)

Reply:
> Docs on the full structure [here](https://github.com/containers/image/blob/8c5dd5b6e7c0253cf2f0f30c8ecd8d5e35cf03c2/docs/containers-signature.5.md#json-data-format), altough I dunno how formal we want to get. I'll work on `oc` tooling for this next week

_Please_ use `c/image/copy.Image`, to do copying with signature verification integrated, if at all possible. Feel free to ping me anytime if you need help with that.

I don’t want copy&pasted partial implementations of that format all over OCP if I can help it.